presto-main/src/main/java/com/facebook/presto/execution/scheduler/nodeSelection/SimpleNodeSelector.java

```diff
@@ -163,10 +159,6 @@ public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTas
                     preferredNodeCount = OptionalInt.of(candidateNodes.size());
                     break;
                 case SOFT_AFFINITY:
-                    // Using all nodes for soft affinity scheduling with modular hashing because otherwise temporarily down nodes would trigger too much rehashing
-                    if (nodeSelectionHashStrategy == MODULAR_HASHING) {
-                        nodeProvider = new ModularHashingNodeProvider(nodeMap.getAllNodes());
-                    }
```

Comment:
Is it intentional to use `nodeMap.allNodes` to build `ModularHashingNodeProvider` for all split nodeSelectionStrategies including `HARD_AFFINITY` and `NO_PREFERENCE` besides `SOFT_AFFINITY`? As I understand the whole change would affect them as well, is that OK?

Reply:
Great question.

`nodeProvider` is only used when there is an affinity preference (`HARD_AFFINITY`, `SOFT_AFFINITY`). When there is no affinity preference (`NO_PREFERENCE`) the `nodeProvider` is not used and random nodes are selected via `randomNodeSelection.pickNodes(split)`.

For `SOFT_AFFINITY` when the `MODULAR_HASHING` strategy is selected nothing changes.

The only place when a potential change is possible is when `HARD_AFFINITY` is enabled, however I don't think it is the case in practice.

When `HARD_AFFINITY` is enabled the connector should return a set of "hard" allocated nodes. Meaning that a split is only allowed to be scheduled on that very specific set of nodes. Hence in theory a connector should never consult a `nodeProvider` when `HARD_AFFINITY` is required, as the `nodeProvider` does not guarantee any specific nodes to be selected. Based on what I checked it's always the case. All connectors that call `NodeProvider#get` only do so when `SOFT_AFFINITY` is preferred:

- https://github.com/prestodb/presto/blob/ff5282e221b461d4e9f9662c0f264cba09211f84/presto-hudi/src/main/java/com/facebook/presto/hudi/HudiSplit.java#L112
- https://github.com/prestodb/presto/blob/ff5282e221b461d4e9f9662c0f264cba09211f84/presto-delta/src/main/java/com/facebook/presto/delta/DeltaSplit.java#L133
- https://github.com/prestodb/presto/blob/ff5282e221b461d4e9f9662c0f264cba09211f84/presto-hive/src/main/java/com/facebook/presto/hive/HiveSplit.java#L170
- https://github.com/prestodb/presto/blob/ff5282e221b461d4e9f9662c0f264cba09211f84/presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergSplit.java#L146

However I agree, the interface today makes it very obscure. I guess a better interface could've been a simple `ConnectorSplit#getSoftAffinityNodeSelectionKey`. We might want to consider a follow up refactor changing the interface to make it more explicit and easier to follow.

Reply:
Thank you for the detailed message. Now I get that in practice, the situation is indeed as you said. And I agree with you that a follow-up refactor to make the interface more explicit would be great. LGTM, thanks!